### PR TITLE
Show tooltip only on overflow

### PR DESCRIPTION
**Pull request by cschleiden**
Exploration 2 of #1433, don't focus on details yet, difference between exploration 1 and 2 is in Tooltiphost

No changefile, will update once we decide on a direction

### CODE REVIEW

File: packages/utilities/src/overflow.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+/**
+ * Check whether an element's content is overflowing
+ *
+ * @param element Element to check for overflow
+ * @returns True if element's content overflows
+ */
+export function isElementOverflowing(element: HTMLElement): boolean {
+  return element.clientWidth < element.scrollWidth;
+}
```
Comment (mdahamiwal) [inline packages/utilities/src/overflow.ts]: we are checking for horizontal overflow. Should be name it accordingly ?

File: packages/office-ui-fabric-react/src/components/Tooltip/TooltipHost.tsx [hunk 2]
```diff
@@ -65,6 +93,21 @@ export class TooltipHost extends BaseComponent<ITooltipHostProps, ITooltipHostSt
...
+  private _checkForOverflow() {
+    const { onlyShowIfOverflow, overflowElement } = this.props;
+    if (onlyShowIfOverflow) {
+      const { isOverflowing } = this.state;
+      const isNowOverflowing = overflowElement && isElementOverflowing(overflowElement);
+
+      if (isNowOverflowing !== isOverflowing) {
+        this.setState({
+          isOverflowing: isNowOverflowing
+        });
```
Comment (mdahamiwal) [inline packages/office-ui-fabric-react/src/components/Tooltip/TooltipHost.tsx]: Any reason why we are keeping "isOverflowing" as a state? how about checking for overflow only on mouseEnter? Seems like we are doing a lot of computation during react lifecycle which may be triggered by parent's state change and produce same result for isOverflowing.

File: packages/utilities/src/overflow.ts (hunk 1 above)
Comment (cschleiden) [inline packages/utilities/src/overflow.ts]: Yep, see Exploration 1 :)

**Comment by cschleiden** [comment]
Desired use case: 
image

**Comment by cschleiden** [comment]
@dzearing @micahgodbolt want to bubble this up again. Reaching out of the component seems a bit weird, but for ease of use reasons I don't see a better way. We could pull this into a different component, or leave it in `TooltipHost`, don't feel very strongly either way.

**Comment by micahgodbolt** [comment]
@cschleiden yeah, this doesn't feel quite right. Mostly because we'll have a ton of other scenarios where we might want to conditionally show a tooltip, and we wouldn't want to cram them into this component. 

I like your overflow utility. Would it be easy enough to make this overflow state something we could easily put on a label, or whatever other component might use it, then maybe have a flag on tooltip that would suppress the tooltip unless the label was overflowed

TL:DR

1. Add "suppress" to the tooltip props
2. Add "overflowed" state/decorator to items that we have overflowing text. 

This way we could do many things (including showing a tooltip) if the text is overflowing.

**Comment by cschleiden** [comment]
@micahgodbolt I don't quite understand. How would I mark an element as containing overflow? I don't want to force people to do something like this:

## Before: Example problem scenario:

### Some shared component:

```CSS
.gridCell {
  overflow: hidden;
  white-space: no-wrap;
  text-overflow: ellipsis; 
}
```
```JSX
<div className="gridCell">
  {content}
</div>
```

Usage of that shared component:
```JSX
const renderCell = () => <span>My text goes here</span>;
```

## After: Example problem scenario:

### Some shared component:

```CSS
.gridCell {
  overflow: hidden;
  white-space: no-wrap;
  text-overflow: ellipsis; 
}
```
```JSX
<OverflowComponent className="gridCell" tagName="div">
  {content}
</OverflowComponent>
```

Usage of that shared component:
```JSX
const renderCell = () => <TooltipHost content="My tooltip goes here" suppressIfNoOverflow={ true }><span>My text goes here</span></TooltipHost>;
```

where `OverflowComponent` would use `context` for example to communicate with `TooltipHost`.. I think that's too heavy. 


______________


I mean if we decide that this is not a problem for other teams, I can contribute the `overflow` utils for now (I think these a not controversial) and keep an `OverflowTooltipHost` in VSTS for our use case. It's a tooltip, it shouldn't take re-architecting a whole control/page to include it :)

**Comment by cschleiden** [comment]
Okay, we could do something like https://github.com/OfficeDev/office-ui-fabric-react/blob/99589edbaff0a68e9c3c6097a3b7be8d058d7c18/packages/utilities/src/scroll.ts#L55 where we set a `data-has-overflow` attribute, that wouldn't require `context`, but still a nice way to attach that element to a parent. 

Also, broke out the overflow changes into https://github.com/OfficeDev/office-ui-fabric-react/pull/1509

**Comment by cschleiden** [comment]
Updated for latest changes. 

Alternative might be to fork `TooltipHost` for a `OverflowTooltipHost`?

**Comment by cschleiden** [comment]
@hross This is the PR I just mentioned.

**Comment by cschleiden** [comment]
@micahgodbolt @dzearing @hross I'd like to move forward with this as soon as possible, we really need that in VSTS.

**Comment by micahgodbolt** [comment]
I have no major objections.

**Comment by hross** [comment]
Per our convo I'd definitely like a delegate to override. I think parent is okay...

enum OverflowMode{ None, Self, Parent } 

overflowBehavior: OverflowMode | (tooltipHostElement: Element) => boolean;

**Comment by cschleiden** [comment]
Now we just need a better way for the `overflowBehavior` property:
```
`enum OverflowMode{ None, Self, Parent }
overflowBehavior: OverflowMode | (tooltipHostElement: Element) => boolean;
```

**Comment by cschleiden** [comment]
Busy remainder of the week, will work on this next week.

**Comment by cschleiden** [comment]
Closing for now, will reopen next week after exploring the `overflowBehavior` a bit

**Comment by cschleiden** [comment]
@hross if you want to have another look
